[DRAFT] Frontends as identifiers in Service REST API

joamaki opened the pull request:
This replaces the use of numerical service identifiers in the service API
and cilium utility with the use of frontend addresses (<ip>:<port>:<protocol>(:<scope>)).

Usage example:
```
    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.0.2:80:TCP \
                                      --backends 172.29.0.3:80 --k8s-node-port
    Creating new service "10.0.0.2:80:TCP"
    Added service with 1 backends
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
    
    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.1.2:80:TCP \
                                      --backends 172.29.1.3:80,172.29.1.4:80 --k8s-node-port
    Creating new service "10.0.1.2:80:TCP"
    Added service with 2 backends
    
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
    10.0.1.2:80:TCP   NodePort       1 => 172.29.1.3:80 (active)
                                     2 => 172.29.1.4:80 (active)
    
    root@838eb2c16285:/home/cilium# cilium service delete 10.0.1.2:80:TCP
    Service "10.0.1.2:80:TCP" deleted successfully
    
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
```

```release-note
The "cilium service" command now uses frontend address, port and protocol tuple as service identifier instead of the internal numerical identifiers. E.g. instead of "cilium service update --id 123 ..." one now does "cilium service update --frontend 10.0.1.2:80:TCP ...". This change allows Cilium to rearrange the identifiers due to e.g. IP address changes without affecting the users who manipulate the services manually.
```

joamaki committed:
cilium: Replace use of integer ids with frontend address in service API

This replaces the use of numerical service identifiers in the service API
and cilium utility with the use of frontend addresses (<ip>:<port>:<protocol>(:<scope>)).

Usage example:

    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.0.2:80:TCP \
                                      --backends 172.29.0.3:80 --k8s-node-port
    Creating new service "10.0.0.2:80:TCP"
    Added service with 1 backends
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)

    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.1.2:80:TCP \
                                      --backends 172.29.1.3:80,172.29.1.4:80 --k8s-node-port
    Creating new service "10.0.1.2:80:TCP"
    Added service with 2 backends

    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
    10.0.1.2:80:TCP   NodePort       1 => 172.29.1.3:80 (active)
                                     2 => 172.29.1.4:80 (active)

    root@838eb2c16285:/home/cilium# cilium service delete 10.0.1.2:80:TCP
    Service "10.0.1.2:80:TCP" deleted successfully

    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)

Signed-off-by: Jussi Maki <jussi@isovalent.com>

oblazek commented on cilium/cmd/service_update.go:
```diff
@@ -92,37 +66,40 @@ func boolToInt(set bool) int {
-	fa := parseFrontendAddress(frontend)
 	skipFrontendCheck := false
 
+	fa, err := loadbalancer.NewL3n4AddrFromModelID(frontend)
```
This breaks the possibility to update backend state(s)... Usually I do `cilium service update --backends 10.249.97.30:8888 --states maintenance` which means frontend is empty string, but inside the `NewL3n4AddrFromModelID` the frontend format is checked -> i.e. `Error: Cannot update service: invalid service frontend "", expected "<ip>:<port>:<proto>(:i)" e.g. 1.2.3.4:80:TCP (IPv4, external scope) or [2001::1]:443:TCP:i (IPv6, internal scope)`

joamaki replied:
Thanks for catching this. Will fix this.

joamaki replied:
Refactored this now. Could you take another look?

joamaki committed:
test/l4lb: Build objects with cilium-builder and expand test

To support testing locally and in systems that don't have suitable clang,
use the cilium-builder image to build the BPF objects for test/l4lb.

Use "docker load" and "docker save" to transfer the cilium-agent image from
the host to lb-node to support local images.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
